torch/csrc/jit/codegen/cuda/lower_magic_zero.cpp

```diff
@@ -88,6 +90,146 @@ bool isProtectedWithMagicZero(const Val* val) {
+  // Find magic zero insertion point
+  IterDomain* magic_zero_loop = nullptr;
+
+  // Search for proper magic zero insertion point,
+  //  prefer innermost.
+  for (auto idx : c10::irange(loops.size())) {
+    auto loop = loops[idx];
+    auto concrete_loop_id = GpuLower::current()->caMap()->getConcreteMappedID(
+        loop_domains[idx], IdMappingMode::EXACT);
+    auto loop_ind = concrete_loop_idx_map.at(concrete_loop_id);
+
+    // Save the concrete id if this loop id is decided to
+    //  be the insertion point by the magic zero util.
+    if (needsMagicZero(loop, concrete_loop_id, loop_ind)) {
+      magic_zero_loop = concrete_loop_id;
+    }
+  }
+
+  // Insert magic zero if insertion point found
+  if (magic_zero_loop != nullptr &&
+      concrete_loop_idx_map.count(magic_zero_loop)) {
+    auto& ind = concrete_loop_idx_map.at(magic_zero_loop);
+    ind = SimplifyingIrBuilder::addExpr(
+        ind, GpuLower::current()->kernel()->magicZeroVal());
+  }
+}
+
+namespace {
+
+//! Protect loop_index_to_protect appearing in overall_index_val
+IndexMagicZeroInfo protectIndexByReplacingLoopIndex(
+    IterDomain* loop_id,
+    Val* overall_index_val,
+    Val* loop_index_to_protect) {
+  auto protected_loop_index = SimplifyingIrBuilder::addExpr(
+      loop_index_to_protect, GpuLower::current()->kernel()->magicZeroVal());
+
+  std::unordered_map<Val*, Val*> replacement_map;
+  replacement_map[loop_index_to_protect] = protected_loop_index;
+
+  auto protected_index =
+      ir_utils::replaceValInIndexVal(overall_index_val, replacement_map);
+
+  IndexMagicZeroInfo info;
+  info.index = protected_index;
+  info.original_loop_index = loop_index_to_protect;
+  info.protected_loop_index = protected_loop_index;
+  info.loop_id = loop_id;
+  return info;
+}
+
+} // namespace
+
+IndexMagicZeroInfo protectPredicateIndexWithMagicZero(
+    Val* index,
+    const IndexFromIdGraph& id_graph,
+    const std::vector<kir::ForLoop*>& loops) {
+  // Gather the loop indices
+  std::unordered_set<Val*> loop_indices;
+  for (auto loop_id : id_graph.resolved_loop_domains) {
+    auto concrete_loop_id = GpuLower::current()->caMap()->getConcreteMappedID(
+        loop_id, IdMappingMode::EXACT);
+    auto index_it = id_graph.initial_concrete_index_map.find(concrete_loop_id);
+    TORCH_INTERNAL_ASSERT(
+        index_it != id_graph.initial_concrete_index_map.end(),
+        "Index not found for loop: ",
+        concrete_loop_id->toString());
+    TORCH_INTERNAL_ASSERT(
+        index_it != id_graph.initial_concrete_index_map.end());
+    auto loop_index = index_it->second;
+    loop_indices.insert(loop_index);
+  }
+
+  // Figure out which loop indices are used in index
+  const auto vals = DependencyCheck::getAllValsBetween(loop_indices, {index});
+
+  // Traverser from the inner-most loop and apply the magic-zero
+  // prorection if needed
+  for (auto loop_id_it = id_graph.resolved_loop_domains.rbegin();
```

Comment:
Just a thought, maybe this loop could just iterate from the right of the loops vector and the corresponding entries on resolved_loop_domains should be permissively mapped. 

vvv Might not be exactly changed this way but just to illustrate. 
```suggestion
  for (auto i=loops.size()-1; i>=0;i--){
     auto loop = loops[i];
     auto loop_id = id_graph.resolved_loop_domains[i];
     TORCH_INTERNAL_ASSERT(
       GpuLower::current()->caMap()->areMapped(
              loop_id, loop->iter_domain(), IdMappingMode::PERMISSIVE));
```

Reply:
Is it guaranteed that `resolved_loop_domains.size() == loops.size()`? I still need to learn about the new interface, but does `resolved_loop_domains` include loops that surround but are not used by an index?

Reply:
I think the answer is yes:
https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/lower_index_compute.cpp#L735-L758

Reply:
Yes, so `resolved_loop_domains.size() >= loops.size()`. It is currently still the same behavior as the reference domain. So 
1. `resolved_loop_domain[0...loops.size()]` should permissively map with iterdomain of loops with the same index.
2. The domains on the right of `loops.size()-1` are dangling domains that we don't expect to show up on the index math.

Reply:
Do you mean the domains in `resolved_loop_domains` are not ordered as the actual loop ordering?

Reply:
They're ordered the same if you look from the left. `resolved_loop_domains` just have additional unused domains on the right of all the actual loops.

Reply:
I thought `loops` here just represents a loop nest, and I thought it's possible that some of the loops aren't used in indexing certain tensors. Am I right?

If `resolved_loop_domains[0...loops.size()] are guaranteed to map with `loops`, it seems that `resolved_loop_domains` also just the loop nest at the time of indexing the given tensor. But what about the remaining domains of `resolved_loop_domains`?  If they are `dangling domains that we don't expect to show up on the index math.`, that seems to indicate `resolved_loop_domains` and `loops` only have used (non-dangling) domains, which contracts with my understanding about `loops`.

Reply:
Yes I think your understanding is correct and I agree it's counter intuitive. The presence of dangling domain is part of the remaining logic from reference tensor and I guess I could get rid of them in a few more cleanup steps.

Reply:
Do you mean the `loops`  vector also has dangling domains? If so, why `resolved_loop_domains.size() >= loops.size()`?

Reply:
No. `Loops` doesn't have dangling nodes. It's the dangling nodes on `resolved_loop_domains` I was referring to.

Reply:
i.e. the vector your looking for is `{resolved_loop_domains.begin(), resolved_loop_domains.begin() + loops.size()}`

Reply:
I inserted `TORCH_INTERNAL_ASSERT(loops.size() == id_graph.resolved_loop_domains.size());` at this point. Looks like no C++ test fails. Is it just by luck?

Reply:
Looks like we're either losing test coverage or the dangling nodes are already removed from the updated replay. I will take a deeper look into this.

Reply:
Updated the code (and added the assertion)

Reply:
Did some archaeology on previous implementation. I think this part of the IdGraph logic was removing a lot of dangling nodes. 
[https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/compute_at_map.cpp#L517-L521](https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/compute_at_map.cpp#L517-L521)

So the loop iterdomains is now picked from a TensorView that's as far down the use chain as possible, which earlier implementation, when this was still called `ParallelMap`, didn't do.

The most common dangling Id I saw from a previous implementation is like this one: (after a rfactoring schedule)
```
T1[I3, I4rf, R5] = sum(T0[I0, I1]) (with I4rf, R5 = split I1)
T2[I6, R7] = sum(T1[I3, I4rf])
```
which **used to** lower to:

```cpp
for x in 0...I3.size:          // Loop0
  // first reduction
  for y in 0...I4rf.size:     // Loop1
    for z in 0...R5.size:    // Loop2
      I1[x,y] += ...

  // second reduction
  for y in 0...I4rf.size:     // Loop3
     T1[x] + = ...
```

In previous implementation `Loop3` would take `I4rf` as its iterdomain which would make reference replay pick up a replayed version of `R5` when traversing up the rfactor domain path, and `R5` would be a dangling node when indexing the second reduction. 

In the current implementation IdGraph picked up `R7` for `Loop3` and in this case `R5` would not show up.

I'm not yet sure if that's the whole story as both the indexing path and the mapping logic have been so different. 

Created #1850 to track further investigation, and could motivate some follow up PRs.

Thanks for pointing this out.

Reply:
Thanks for looking into it. Looks like either way seems to work fine.